File: 525-contiguous-array/525-contiguous-array.scala
import scala.collection.mutable._
        
object Solution {
    def findMaxLength(nums: Array[Int]): Int = {
        val map: Map[Int, Int] = Map()
        map += (0 -> -1)

        var cnt = 0
        var res = 0

        nums.zipWithIndex.foreach { case (num, idx) =>
            if (num == 0) cnt -= 1 else cnt += 1
            if (map.contains(cnt)) {
                res = res max idx - map.get(cnt).get
            } else {
                map += (cnt -> idx)
            }
        }

        res 
    }
}

File: 525-contiguous-array/525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        int res = 0, balance = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(balance, -1);
        
        for(int i = 0; i < nums.length; i++) {
            int n = nums[i];
            balance += (n == 0)? -1 : 1;
            if (!map.containsKey(balance)) map.put(balance, i);
            res = Math.max(res, i - map.get(balance));
        }
        
        return res;
    }
}